### DISCUSSION TITLE
Running Multithreaded Zephyr with Micropython

### DISCUSSION THREAD

**Discussion by jandriea** [micropython/micropython#13265] [Zephyr]
Hello everyone,

I am currently working on a project that requires a C/C++-based app combined with a Micropython app. Initially, my thought was to include a single function call within the main function. However, I believe it would be more efficient to implement an RTOS and create a distinct task for both apps.

Here's a high-level overview of the project structure:
- RTOS (Zephyr) servers as the operating system foundation.
- Thread A is dedicated to running App A in a standard fashion.
- Thread B is tasked with running Micropython

For context, we already have a significant codebase in Micropython, and we are keen on reusing this existing code. However, we also require the inclusion of a real-time application that operates in a separate task with a higher priority.

If anyone has experience with a similar multithreaded Zephyr and Micropython setup or can provide guidance on seamlessly integrating these components, I would greatly appreciate your input.

Thank you in advance for your time and assistance!

**Comment by peterhinch** [comment]
You haven't mentioned your hardware platform, but there might be a bare metal alternative if you considered RP2. MicroPython would run on core 1, with your C code running on core 2. That would give you true concurrency with no time-slicing.

**Reply by jandriea** [reply]
Thank you for your reply,

I appreciate your suggestion regarding the RP2 platform and the potential for achieving true concurrency with MicroPython on core 1 and C code on core 2. However, our hardware platform of choice is a custom board featuring the STM32U575.

I have tried to compile and run the multithread Zephyr RTOS with one thread running the minimal MicroPython REPL and the other thread running something else. However, when I attempt to add a loop inside the MicroPython REPL, the other thread stops running.

**Comment by chrhoehl** [comment]
Hi, @jandriea  I am thinking of a similar approach to make use of the Zephyr real time capabilities and hardware abstraction while implementing high level program flow in micropython.

Did you ever find a solution?

**Reply by jandriea** [reply]
Sadly, I still haven't found any solution for this.
